### PR TITLE
Fix permissions complexity limits

### PR DESCRIPTION
**Pull request by Neopallium**
Calculates the complexity of permissions for secondary keys.  Each identity has a budget limit for the total complexity of call secondary key permissions.

### modified logic

- More restrict limits on secondary keys and their permissions.

### CODE REVIEW

File: primitives/src/identity.rs [hunk 1]
```diff
@@ -69,6 +69,13 @@ where
         });
         self
     }
+
+    /// Returns the complexity of the permissions.
```
Comment (Centril) [inline primitives/src/identity.rs:73]:
```suggestion
    /// Returns the complexity of the DID's permissions.
```

File: primitives/src/secondary_key.rs [hunk 2]
```diff
@@ -293,6 +323,11 @@ where
     pub fn has_portfolio_permission(&self, it: impl IntoIterator<Item = PortfolioId>) -> bool {
         self.permissions.portfolio.ge(&SubsetRestriction::elems(it))
     }
+
+    /// Returns the complexity of the permissions.
```
Comment (Centril) [inline primitives/src/secondary_key.rs:327]:
```suggestion
    /// Returns the complexity of the secondary key's permissions.
```

File: primitives/src/secondary_key.rs [hunk 3]
```diff
@@ -142,6 +154,24 @@ impl Permissions {
...
         ));
     }
+
+    /// Returns the complexity of the permissions.
+    pub fn complexity(&self) -> usize {
+        // Calculate the pallet/extrinsic permissions complexity cost.
+        let cost = self.extrinsic.inner().map_or(0usize, |set| {
+            set.iter()
+                .fold(0, |cost, pallet| cost.saturating_add(pallet.complexity()))
+        });
```
Comment (Centril) [inline primitives/src/secondary_key.rs:164]: Can you add `fn fold<B>(&self, A, impl FnMut(B, &A) -> B) -> B, ` on `SubsetRestriction` and refactor the two cases here?

File: pallets/identity/src/keys.rs [hunk 4]
```diff
@@ -494,14 +508,25 @@ impl<T: Config> Module<T> {
...
+        Self::ensure_secondary_keys_limited(&record, 1, perms.complexity())?;
 
         Self::ensure_key_did_unlinked(&key)?;
         Ok(())
     }
 
+    pub fn ensure_secondary_keys_limited(
+        record: &DidRecord<T::AccountId>,
+        keys: usize,
+        cost: usize,
```
Comment (Centril) [inline pallets/identity/src/keys.rs:523]: Could do with a doc comment explaining `keys` & `cost`.

**Comment by Centril** [comment]
Looks good functionally. Some nits ^--

**Review by Centril — APPROVED** [review]

**Review by Centril — APPROVED** [review]

**Review by adamdossa — APPROVED** [review]

**Pull request merged**